[precompile] Fix issues with guard serialization on distributed types.

zhxchen17 opened the pull request:
Summary: Add more support for torch internal distributed data structures.

Test Plan:
test_guard_serialization.py

Rollback Plan:

Differential Revision: D81927732




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

zhxchen17 commented on test/dynamo/test_guard_serialization.py:
```diff
@@ -1346,6 +1355,49 @@ def forward(self, x):
+
+        mesh = init_device_mesh(str(torch.get_default_device()), (1,))
+        m = SimpleModule()
+        m = fully_shard(m, mesh=mesh)
```
@ezyang oh this is for caching precompile, not the aot/fullgraph case. James and I are working on an orthogonal use case (omnifm) and it uses fsdp on some module.

dolpm commented on torch/_dynamo/guards.py:
```diff
@@ -3222,6 +3252,9 @@ def reducer_override(
         elif isinstance(obj, types.MappingProxyType):
             return type(self)._unpickle_mapping_proxy, (obj.copy(),)
 
+        elif isinstance(obj, type({}.keys())):
```
nit: use https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/utils.py#L2567 ?

dolpm commented on torch/_dynamo/guards.py:
```diff
@@ -3234,14 +3267,27 @@ def reducer_override(
+                types.CodeType,
+                torch._C.Stream,
+                torch._C._distributed_c10d.ProcessGroup,
+                weakref.ReferenceType,
```
where do we get this list?

zhxchen17 replied:
we just ran precompile through an internal model with some complex data structure and saw these not needed but serialized.

zhxchen17 committed:
[precompile] Fix issues with guard serialization on distributed types. (#162418)

Summary:

Add more support for torch internal distributed data structures.

Test Plan:
test_guard_serialization.py

Rollback Plan:

Reviewed By: dolpm

Differential Revision: D81927732
